### PR TITLE
[DEVC-686] Reject NTRIP settings while NTRIP is enabled

### PR DESCRIPTION
**Pull request by pmiettinen**
Fix for https://swift-nav.atlassian.net/browse/DEVC-686

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-8-g70d1f920
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=661059)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-8-g70d1f920&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-8-g70d1f920&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-8-g70d1f920
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-8-g70d1f920)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-8-g70d1f920/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-8-g70d1f920

File: package/ntrip_daemon/src/ntrip_settings.c [hunk 1]
```diff
@@ -150,24 +151,57 @@ void ntrip_stop_processes()
   }
 }
 
+/* Move to libpiksi as generic function */
+static bool file_read_value(char *file_path) {
```
Comment (benjaminaltieri) [inline package/ntrip_daemon/src/ntrip_settings.c]: Could un-static this function in libpiksi: https://github.com/swift-nav/piksi_buildroot/blob/84ee95fe09f9196c49fee0a0306cea23601f5683/package/libpiksi/libpiksi/src/util.c#L34

File: package/ntrip_daemon/src/ntrip_settings.c [hunk 2]
```diff
@@ -150,24 +151,57 @@ void ntrip_stop_processes()
...
+}
+
 static int ntrip_notify(void *context)
 {
   (void)context;
 
+  /* If NTRIP was enabled before notify call and ntrip_enabled is still true,
+   * it means that user is trying to modify some other NTRIP related setting
+   * than ntrip_enabled -> reject and instruct. */
+  if (file_read_value(NTRIP_ENABLED_FILE_PATH) && ntrip_enabled) {
```
Comment (benjaminaltieri) [inline package/ntrip_daemon/src/ntrip_settings.c]: I'm actually not quite sure I understand why we need to read this file as the ntrip settings loop should be the only modifier of the value. Should be ok to just check `ntrip_enabled`

File: package/ntrip_daemon/src/ntrip_settings.c [hunk 3]
```diff
@@ -150,24 +151,57 @@ void ntrip_stop_processes()
...
+
   snprintf(ntrip_interval_s, sizeof(ntrip_interval_s), "%d", ntrip_interval);
   ntrip_rev1gga_s[0] = ntrip_rev1gga ? 'y' : 'n';
 
   for (size_t i = 0; i < ntrip_processes_count; i++) {
 
     ntrip_stop_process(i);
 
-    // TODO: Remove the need for these by reading from control socket
+    /* TODO: Remove the need for these by reading from control socket */
```
Comment (benjaminaltieri) [inline package/ntrip_daemon/src/ntrip_settings.c]: @silverjam is this TODO still relevant? maybe @pmiettinen can implement while he's touching this code

Comment (silverjam) [reply package/ntrip_daemon/src/ntrip_settings.c]: TODO is still relevant, but probably not germane for this PR, we could add a ticket for it though, basically want to re-use to the command socket that's used for `ntrip_daemon --reconnect` to make an `ntrip_daemon --is-enabled` so we don't have maintain these files.

File: package/ntrip_daemon/src/ntrip_settings.c (hunk 2 above)
Comment (pmiettinen) [reply package/ntrip_daemon/src/ntrip_settings.c]: Could be that I have understood wrongly how the settings framework behaves but let me try to explain how I see it with my current understanding.

There is one case why this is necessary: If user is changing `ntrip.enable` value `False` -> `True`. My assumption is that settings framework first reads the value `True` into `ntrip_enabled` and only after that `ntrip_notify` is called. Now if we only observe `ntrip_enabled` we read the value `True`, issue this new warning and the worst: don't actually do the enabling actions.

Do you think this is correct understanding?

Comment (pmiettinen) [reply package/ntrip_daemon/src/ntrip_settings.c]: I'll test with `if (ntrip_enabled)` and see the results support my explanation.

Comment (pmiettinen) [reply package/ntrip_daemon/src/ntrip_settings.c]: Test result supports my understanding. `if (ntrip_enabled)` doesn't let me change `ntrip.enable` value `False` -> `True`.

Comment (pmiettinen) [reply package/ntrip_daemon/src/ntrip_settings.c]: Ended up reworking the logic due to other issues. See relevant next comments.

File: package/libpiksi/libpiksi/include/libpiksi/util.h [hunk 4]
```diff
@@ -77,6 +77,8 @@ bool device_has_gps_time(void);
  */
 void set_device_has_gps_time(bool has_time);
 
+bool file_read_value(char *file_path);
```
Comment (pmiettinen) [inline package/libpiksi/libpiksi/include/libpiksi/util.h]: Looks like this isn't needed after all in this PR but I think it's a good function anyway so I propose to merge it anyway.

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-12-gd2d3d7ec
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=662440)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-12-gd2d3d7ec&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-12-gd2d3d7ec&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-12-gd2d3d7ec
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-12-gd2d3d7ec)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-12-gd2d3d7ec/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-12-gd2d3d7ec

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-13-g336d2e9a
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=662496)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-13-g336d2e9a&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-13-g336d2e9a&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-13-g336d2e9a
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-13-g336d2e9a)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-13-g336d2e9a/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-13-g336d2e9a

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-14-gdc5cdae0
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=662556)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-14-gdc5cdae0&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-14-gdc5cdae0&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-14-gdc5cdae0
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-14-gdc5cdae0)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-14-gdc5cdae0/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-14-gdc5cdae0

**Comment by anonymous-travis-swiftnav** [comment]
## HITL smoke tests: v1.5.0-develop-2018083120-14-gdc5cdae0
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=662587)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-14-gdc5cdae0&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v1.5.0-develop-2018083120-14-gdc5cdae0&groupby_key=firmware&display_type=table)

**Comment by anonymous-travis-swiftnav** [comment]
## v1.5.0-develop-2018083120-14-gdc5cdae0
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v1.5.0-develop-2018083120-14-gdc5cdae0)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.5.0-develop-2018083120-14-gdc5cdae0/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.5.0-develop-2018083120-14-gdc5cdae0

**Comment by silverjam** [comment]
:shipit:

**Pull request merged**